docs/rfc/rfc-008-don't-panic.md

```diff
@@ -0,0 +1,133 @@
+
+### References
+
+- [Defer Panic and Recover](https://go.dev/blog/defer-panic-and-recover)
+- [Why Go gets exceptions right](https://dave.cheney.net/tag/panic)
+- [Don't panic](https://dave.cheney.net/practical-go/presentations/gophercon-singapore-2019.html#_dont_panic)
+
+## Discussion
+
+### Acceptable Panics
+
+#### Initialization
+
+It is unambiguously safe (and desireable) to panic in `init()`
+functions in response to any kind of error. These errors are caught by
+tests, and occur early enough in process initialization that they
+won't cause unexpected runtime crashes.
+
+Other code that is called early in process initialization MAY panic,
+in some situations if it's not possible to return an error or cause
+the process to abort early, although these situations should be
+vanishingly slim.
+
+#### Data Corruption
+
+If Tendermint code encounters an inconsistency that could be
+attributed to data corruption or a logical impossibility it is safer
+to panic and crash the process than continue to attempt to make
+progress in these situations.
+
+Examples including reading data out of the storage engine that
+is invalid or corrupt, or encountering an ambiguous situation where
+the process should halt. Generally these forms of corruption are
+detected after interacting with a trusted but external data source,
+and reflect situations where the author thinks its safer to terminate
+the process immediately rather than allow execution to continue.
+
+TODO: Add more and better examples here.
+
+#### Static Validity
+
+It is acceptable to panic for invariant violations, within a library
+or package, in situations that should be statically impossible,
+because there is no way to make these kinds of assertions at compile
+time.
+
+For example, type-asserting `interface{}` values returned by
+`container/list` and `container/heap` (and similar), is acceptable,
+because package authors should have exclusive control of the inputs to
+these containers. Packages should not expose the ability to add
+arbitrary values to these data structures.
+
+#### Controlled Panics Within Libraries
+
+In some algorithms with highly recursive structures or very nested
+call patterns, using a panic, in combination with conditional recovery
+handlers results in more manageable code. Ultimately this is a limited
+application, and implementations that use panics internally should
+only recover conditionally, filtering out panics rather than ignoring
+or handling all panics.
+
+#### Request Handling
+
+Code that handles responses to incoming/external requests
+(e.g. `http.Handler`) should avoid panics, but practice this isn't
+totally possible, and it makes sense that request handlers have some
+kind of default recovery mechanism that will prevent one request from
+terminating a service.
+
+### Unacceptable Panics
+
+In **no** other situation is it acceptable for the code to panic:
+
+- there should be controlled panics that should be handled across
```

Comment:
I'm a little unclear, are you saying that this is a case where panics should not exist? I generally think that we should endeavor to avoid panics in this case and not ever expect to handle a panic for a package we called into. If the package is expected to shutdown in error, it should communicate that with an error.

Reply:
I think I forgot some negation words here. I've cleaned that up.